src/wallet/rpc/wallet.cpp

```diff
@@ -817,6 +817,113 @@ static RPCHelpMan migratewallet()
+            HelpExampleCli("gethdkeys", "") + HelpExampleRpc("gethdkeys", "")
+            + HelpExampleCliNamed("gethdkeys", {{"active_only", "true"}, {"private", "true"}}) + HelpExampleRpcNamed("gethdkeys", {{"active_only", "true"}, {"private", "true"}})
+        },
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+        {
+            const std::shared_ptr<const CWallet> wallet = GetWalletForJSONRPCRequest(request);
+            if (!wallet) return UniValue::VNULL;
+
+            if (!wallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
+                throw JSONRPCError(RPC_WALLET_ERROR, "gethdkeys is not available for non-descriptor wallets");
+            }
+
+            LOCK(wallet->cs_wallet);
+
+            UniValue options{request.params[0].isNull() ? UniValue::VOBJ : request.params[0]};
+            const bool active_only{options.exists("active_only") ? options["active_only"].get_bool() : false};
+            const bool priv{options.exists("private") ? options["private"].get_bool() : false};
+            if (priv) {
+                EnsureWalletIsUnlocked(*wallet);
+            }
+
+
+            std::set<ScriptPubKeyMan*> spkms;
+            if (active_only) {
+                spkms = wallet->GetActiveScriptPubKeyMans();
+            } else {
+                spkms = wallet->GetAllScriptPubKeyMans();
+            }
+
+            std::map<CExtPubKey, std::set<std::tuple<std::string, bool, bool>>> wallet_xpubs;
+            std::map<CExtPubKey, CExtKey> wallet_xprvs;
+            for (auto* spkm : spkms) {
+                auto* desc_spkm{dynamic_cast<DescriptorScriptPubKeyMan*>(spkm)};
+                CHECK_NONFATAL(desc_spkm);
+                LOCK(desc_spkm->cs_desc_man);
+                WalletDescriptor w_desc = desc_spkm->GetWalletDescriptor();
+
+                // Retrieve the pubkeys from the descriptor
+                std::set<CPubKey> desc_pubkeys;
+                std::set<CExtPubKey> desc_xpubs;
+                w_desc.descriptor->GetPubKeys(desc_pubkeys, desc_xpubs);
+                for (const CExtPubKey& xpub : desc_xpubs) {
+                    std::string desc_str;
+                    desc_spkm->GetDescriptorString(desc_str, false);
+                    wallet_xpubs[xpub].emplace(desc_str, wallet->IsActiveScriptPubKeyMan(*spkm), desc_spkm->HasPrivKey(xpub.pubkey.GetID()));
+                    if (std::optional<CKey> key = priv ? desc_spkm->GetKey(xpub.pubkey.GetID()) : std::nullopt) {
+                        wallet_xprvs[xpub] = CExtKey(xpub, *key);
+                    }
+                }
+            }
+
+            UniValue response(UniValue::VARR);
+            for (const auto& [xpub, descs] : wallet_xpubs) {
+                bool has_xprv = false;
+                UniValue descriptors(UniValue::VARR);
+                for (const auto& [desc, active, has_priv] : descs) {
+                    UniValue d(UniValue::VOBJ);
+                    d.pushKV("desc", desc);
```

Comment:
In 21d0bc38ab:

Better to return the descriptors sorted:
```diff
diff --git a/src/wallet/rpc/wallet.cpp b/src/wallet/rpc/wallet.cpp
--- a/src/wallet/rpc/wallet.cpp	(revision d7fc8a8bec5edb657f65ef35042d801e2913ab81)
+++ b/src/wallet/rpc/wallet.cpp	(date 1710943877002)
@@ -874,7 +874,7 @@
                 spkms = wallet->GetAllScriptPubKeyMans();
             }
 
-            std::map<CExtPubKey, std::set<std::tuple<std::string, bool, bool>>> wallet_xpubs;
+            std::map<CExtPubKey, std::vector<std::tuple<std::string, bool, bool>>> wallet_xpubs;
             std::map<CExtPubKey, CExtKey> wallet_xprvs;
             for (auto* spkm : spkms) {
                 auto* desc_spkm{dynamic_cast<DescriptorScriptPubKeyMan*>(spkm)};
@@ -889,7 +889,7 @@
                 for (const CExtPubKey& xpub : desc_xpubs) {
                     std::string desc_str;
                     desc_spkm->GetDescriptorString(desc_str, false);
-                    wallet_xpubs[xpub].emplace(desc_str, wallet->IsActiveScriptPubKeyMan(*spkm), desc_spkm->HasPrivKey(xpub.pubkey.GetID()));
+                    wallet_xpubs[xpub].emplace_back(desc_str, wallet->IsActiveScriptPubKeyMan(*spkm), desc_spkm->HasPrivKey(xpub.pubkey.GetID()));
                     if (std::optional<CKey> key = priv ? desc_spkm->GetKey(xpub.pubkey.GetID()) : std::nullopt) {
                         wallet_xprvs[xpub] = CExtKey(xpub, *key);
                     }
@@ -897,9 +897,15 @@
             }
 
             UniValue response(UniValue::VARR);
-            for (const auto& [xpub, descs] : wallet_xpubs) {
-                bool has_xprv = false;
+            for (auto& [xpub, descs] : wallet_xpubs) {
                 UniValue descriptors(UniValue::VARR);
+
+                // Sort descriptors
+                std::sort(descs.begin(), descs.end(), [](const auto& a, const auto& b) {
+                    return std::get<0>(a) < std::get<0>(b);
+                });
+
+                bool has_xprv = false;
                 for (const auto& [desc, active, has_priv] : descs) {
                     UniValue d(UniValue::VOBJ);
                     d.pushKV("desc", desc);
```

Reply:
I don't think it's meaningfully useful to explicitly sort them, and we don't do that anywhere else. Also, `std::set` does sorting itself, so I don't think this would even change the output.

Reply:
> I don't think it's meaningfully useful to explicitly sort them, and we don't do that anywhere else.

We sort them in `listdescriptors`: [src/wallet/rpc/backup.cpp#L1827](https://github.com/bitcoin/bitcoin/blob/b50554babdddf452acaa51bac757736766c70e81/src/wallet/rpc/backup.cpp#L1827).

> Also, `std::set` does sorting itself, so I don't think this would even change the output.

`std::set` sorts based on the top element, the std::tuple in this case. Not based on the descriptor string.

-------

Yet, this was just an UX nit. Either way is fine for me.

Reply:
Sorting on `std::tuple` will sort lexicographically on the items in the tuple. Since the descriptor string is the first item in the tuple, and are unique, it should be sorted by the descriptor strings.

Reply:
> Sorting on `std::tuple` will sort lexicographically on the items in the tuple. Since the descriptor string is the first item in the tuple, and are unique, it should be sorted by the descriptor strings.

Hmm, I was sure I saw a different ordering this morning. But 🤷🏼 ..

Reply:
Just tried and they're neatly grouped by type. `tr()` is listed before `wpkh` which takes getting used to since we introduced `tr()` later, but such is the alphabet.